src/antlr/QueryStructureVisitor.ts

```diff
@@ -0,0 +1,520 @@
+        if (tableCol !== undefined) {
+          return tableCol.column.name;
+        }
+      }
+    }
+    return undefined;
+  }
+
+  /**
+   *  Determines whether expression is a star.
+   */
+
+  private isStar(ctx: ExpressionContext): StarContext | undefined {
+    const boolExpr = ctx.booleanExpression();
+    if (boolExpr instanceof PredicatedContext) {
+      const valExpr = boolExpr.valueExpression();
+      if (valExpr instanceof ValueExpressionDefaultContext) {
+        const primaryExpr = valExpr.primaryExpression();
+        if (primaryExpr instanceof StarContext) {
+          return primaryExpr;
+        }
+      }
+    }
+    return undefined;
+  }
+
+  /**
+   * Called when relation is ready.
+   * @param _relation
+   * @param _alias
+   * @returns
+   */
+  onRelation(_relation: TableRelation | QueryRelation, _alias?: string): void {
+    return;
+  }
+
+  /**
+   * Called when column reference is ready.
+   * @param _tableId
+   * @param _columnId
+   * @returns
+   */
+  onColumnReference(_tableId: string, _columnId: string): void {
+    return;
+  }
+
+  processStar(ctx: StarContext): Result {
+    const range = this.rangeFromContext(ctx);
+    const qualifiedName = ctx.qualifiedName();
+    if (qualifiedName !== undefined) {
+      // TODO support multipart table names
+      const lastName = qualifiedName.identifier()[qualifiedName.identifier().length - 1];
+      const tableName = common.stripQuote(lastName);
+      const rel = this.currentRelation.findLocalRelation(tableName);
+      if (rel !== undefined) {
+        this.addRelationColumns(rel, range);
+      }
+    } else {
+      for (const r of this.currentRelation.relations) {
+        this.addRelationColumns(r[1], range);
+      }
+    }
+
+    return this.visitChildren(ctx);
+  }
+
+  protected addRelationColumns(rel: Relation, range: Range): void {
+    rel.columns.forEach(c => {
+      const columnId = this.currentRelation.getNextColumnId();
+      const col = new Column(columnId, c.label, range);
+      this.currentRelation.columns.push(col);
+      this.currentRelation.currentColumnId = columnId;
+      this.onColumnReference(rel.id, c.id);
+      this.currentRelation.currentColumnId = undefined;
+    });
+  }
+
+  private processClause(clause: EdgeType, ctx: RuleNode): Result {
+    this.currentRelation.currentClause = clause;
+    const result = this.visitChildren(ctx);
+    this.currentRelation.currentClause = undefined;
+    return result;
+  }
+
+  //
+  // Visitor method overrides
+  //
+
+  visitQuery(ctx: QueryContext): Result {
+    this.currentRelation = new QueryRelation(
+      this.getNextRelationId(),
+      this.currentRelation,
+      this.rangeFromContext(ctx)
+    );
+
+    const result = this.visitChildren(ctx);
+
+    // to be consumed later
+    this.lastRelation = this.currentRelation;
+    if (this.currentRelation.id == "result_1") this.onRelation(this.currentRelation, "[final result]");
```

Comment:
We should maybe extract those `"result_1"` and `"[finalresult]"` somewhere, I tend to avoid hardcoded string in the codebase, this is usually getting out of sync in a very sneaky way 😅